kernel/src/main/java/org/kframework/utils/BinaryLoader.java

```diff
@@ -112,27 +112,30 @@ public void saveOrDie(File file, Object o) {
      * Locks the file before writing, so that it cannot be read by another instance of K. If the file is currently in
      * use, this method will block until lock can be acquired.
      */
-    public void saveSynchronized(FileOutputStream out, Object o) throws IOException, InterruptedException {
+    public void saveSynchronized(File file, Object o) throws IOException, InterruptedException {
         //To protect from concurrent access from another thread
         lock.writeLock().lockInterruptibly();
-        try {
+        //JDK API limitation: there's no API to atomically open a file for writing and lock it.
+        //Consequently, if another thread reads a file between the moments this thread opens a stream and acquires a
+        // lock, it will see an empty file. We use lockStream open for appending as workaround.
+        try (FileOutputStream lockStream = new FileOutputStream(file, true)) {
```

Comment:
this would be less confusing if we just directly got a file channel via FileChannel.open(file.toPath()) rather than creating a file in append mode and then not appending anything.